File: tests/config_example/test_basic_config.cpp
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>
#include <gtest/gtest.h>
#include "tests/config_example/basic_generated.h"
#include <cstddef>
#include <cstdlib>

using namespace config::example;

class ConfigTest : public ::testing::TestWithParam<const char*> {
protected:
    const char *test_file = nullptr;
    void *mapped_data = nullptr;
    size_t file_size = 0;

    void SetUp() override {
        // If CONFIG_FILE is set, always use it for the first file
        if (strcmp(GetParam(), "tests/config_example/etc/basic_test_config.bin") == 0) {
            const char* env_file = getenv("CONFIG_FILE");
            if (env_file) {
                test_file = env_file;
            } else {
                test_file = GetParam();
            }
        } else {
            test_file = GetParam();
        }

        int fd = open(test_file, O_RDONLY);
        ASSERT_NE(fd, -1) << "Failed to open test file: " << test_file;

        struct stat sb;
        ASSERT_EQ(fstat(fd, &sb), 0) << "Failed to get file size.";
        file_size = sb.st_size;

        mapped_data = mmap(nullptr, file_size, PROT_READ, MAP_PRIVATE, fd, 0);
        close(fd);
        ASSERT_NE(mapped_data, MAP_FAILED) << "Failed to mmap file.";
    }

    void TearDown() override {
        if (mapped_data) {
            munmap(mapped_data, file_size);
        }
    }
};

INSTANTIATE_TEST_SUITE_P(
    ConfigFiles,
    ConfigTest,
    ::testing::Values(
        "tests/config_example/etc/basic_test_config.bin",
        "tests/config_example/etc/basic_test_config_evolution.bin"
    )
);

TEST_P(ConfigTest, ValidateAppConfig) {
    auto app_config = GetAppConfig(mapped_data);
    ASSERT_NE(app_config, nullptr);
    EXPECT_EQ(app_config->schema_version()->major(), 1);
    EXPECT_GE(app_config->schema_version()->minor(), 0);
    EXPECT_EQ(app_config->schema_version()->patch(), 0);
    EXPECT_EQ(app_config->app_name()->str(), "TestApp");
    EXPECT_EQ(app_config->app_id(), 0); // Implicit default is taken
    EXPECT_FALSE(app_config->debug_enabled()); // Default value is taken
    EXPECT_EQ(app_config->max_connections(), 100); // Default value is taken
    EXPECT_EQ(app_config->timeout_ms(), 5000); // Default value is taken
}

TEST_P(ConfigTest, ValidateAdvancedSettings) {
    auto app_config = GetAppConfig(mapped_data);
    ASSERT_NE(app_config, nullptr);

    auto advanced_settings = app_config->advanced_settings();
    ASSERT_NE(advanced_settings, nullptr);
    EXPECT_EQ(advanced_settings->log_level()->str(), "INFO");
    EXPECT_EQ(advanced_settings->buffer_size_kb(), 2048);
    EXPECT_TRUE(advanced_settings->enable_metrics());
    ASSERT_EQ(advanced_settings->allowed_hosts()->size(), 2);
    EXPECT_EQ(advanced_settings->allowed_hosts()->Get(0)->str(), "host1");
    EXPECT_EQ(advanced_settings->allowed_hosts()->Get(1)->str(), "host2");
}

TEST(ConfigTest, AccessGarbageBuffer) {
    const size_t buffer_size = 1024;
    uint8_t garbage_buffer[buffer_size];

    // Fill with deterministic garbage data (pattern based on index)
    for (size_t i = 0; i < buffer_size; i++) {
        garbage_buffer[i] = static_cast<uint8_t>((i * 0x5A + 0xAA) & 0xFF);
    }

    // Attempt to access as a FlatBuffer without verification
    auto app_config = GetAppConfig(garbage_buffer);
    EXPECT_NE(app_config, nullptr);
    // Warning! Attempting to access fields will cause undefined behavior

    flatbuffers::Verifier verifier(garbage_buffer, buffer_size);
    EXPECT_FALSE(VerifyAppConfigBuffer(verifier));
}

File: tests/config_example/test_basic_config.rs
use std::fs::File;
use memmap2::Mmap;
use flatbuffers::root;
use lib_basic_rust::config::example::AppConfig;

// Replace with baselibs binary config loader abstraction
fn load_config_file() -> Mmap {
    let file = File::open("tests/config_example/etc/basic_test_config.bin")
        .expect("Failed to open test file");
    unsafe {
        Mmap::map(&file).expect("Failed to mmap file")
    }
}

#[test]
fn validate_app_config() {
    let buffer = load_config_file();
    let app_config = root::<AppConfig>(&buffer).expect("Failed to parse AppConfig");

    assert_eq!(app_config.schema_version().major(), 1);
    assert_eq!(app_config.schema_version().minor(), 0);
    assert_eq!(app_config.schema_version().patch(), 0);
    assert_eq!(app_config.app_name(), "TestApp");
    assert_eq!(app_config.app_id(), 0); // Implicit default is taken
    assert!(!app_config.debug_enabled()); // Default value is taken
    assert_eq!(app_config.max_connections(), 100); // Default value is taken
    assert_eq!(app_config.timeout_ms(), 5000); // Default value is taken
}

#[test]
fn validate_advanced_settings() {
    let buffer = load_config_file();
    let app_config = root::<AppConfig>(&buffer).expect("Failed to parse AppConfig");
    let advanced_settings = app_config.advanced_settings();

    assert_eq!(advanced_settings.log_level(), Some("INFO"));
    assert_eq!(advanced_settings.buffer_size_kb(), 2048);
    assert!(advanced_settings.enable_metrics());
    let allowed_hosts = advanced_settings.allowed_hosts().expect("Failed to get allowed hosts");
    assert_eq!(allowed_hosts.len(), 2);
    assert_eq!(allowed_hosts.get(0), "host1");
    assert_eq!(allowed_hosts.get(1), "host2");
}

#[test]
fn access_garbage_buffer() {
    const BUFFER_SIZE: usize = 1024;
    let mut garbage_buffer = vec![0u8; BUFFER_SIZE];

    // Fill with deterministic garbage data (pattern based on index)
    for i in 0..BUFFER_SIZE {
        garbage_buffer[i] = ((i * 0x5A + 0xAA) & 0xFF) as u8;
    }

    unsafe {
        let _unchecked_root = flatbuffers::root_unchecked::<AppConfig>(&garbage_buffer);
        // Warning! Attempting to access fields will cause undefined behavior
    }

    let verified_root = root::<AppConfig>(&garbage_buffer);
    assert_eq!(verified_root.is_err(), true);
}
